azure/hostedservice/client.go

```diff
@@ -0,0 +1,135 @@
+package hostedservice
+
+import (
+	"encoding/base64"
+	"encoding/xml"
+	"fmt"
+
+	"github.com/MSOpenTech/azure-sdk-for-go/azure"
+	locationclient "github.com/MSOpenTech/azure-sdk-for-go/azure/location"
+)
+
+const (
+	azureXmlns                        = "http://schemas.microsoft.com/windowsazure"
+	azureDeploymentListURL            = "services/hostedservices/%s/deployments"
+	azureHostedServiceListURL         = "services/hostedservices"
+	deleteAzureHostedServiceURL       = "services/hostedservices/%s?comp=media"
+	azureHostedServiceAvailabilityURL = "services/hostedservices/operations/isavailable/%s"
+	azureDeploymentURL                = "services/hostedservices/%s/deployments/%s"
+	deleteAzureDeploymentURL          = "services/hostedservices/%s/deployments/%s?comp=media"
+
+	errParamNotSpecified = "Parameter %s is not specified."
+	errInvalidDnsLength  = "The DNS name must be between 3 and 25 characters."
+)
+
+//NewClient is used to return a handle to the HostedService API
+func NewClient(client *azure.Client) *HostedServiceClient {
+	return &HostedServiceClient{client: client}
+}
+
+func (self *HostedServiceClient) CreateHostedService(dnsName, location string, reverseDnsFqdn string, label string, description string) (string, error) {
+	if len(dnsName) == 0 {
+		return "", fmt.Errorf(errParamNotSpecified, "dnsName")
+	}
+	if len(location) == 0 {
+		return "", fmt.Errorf(errParamNotSpecified, "location")
+	}
+
+	err := self.verifyDNSName(dnsName)
+	if err != nil {
+		return "", err
+	}
+
+	result, reason, err := self.CheckHostedServiceNameAvailability(dnsName)
+	if err != nil {
+		return "", err
+	}
+	if !result {
+		return "", fmt.Errorf("%s Hosted service name: %s", reason, dnsName)
+	}
+
+	locationClient := locationclient.NewClient(self.client)
```

Comment:
@hopetobelievein the reason was more subtle than that - it's because despite us doing the call, it can still fail on creation with a Conflict - so we may as well just let it fail there. `CheckHostedServiceNameAvailability` is only really useful for user-facing CLI tools, since any other use results in a race condition.